src/storage-types/src/sources.rs

```diff
@@ -248,6 +252,16 @@ pub struct SourceExport<S = ()> {
     /// to adjust the output index, e.g. by adding 1 to the index of the name
     /// they search for.
     pub output_index: usize,
+    /// The width of this subsource's output.
+    ///
+    /// This field allows us to demux one output stream from an ingestion into
+    /// multiple persist shards with (ostensibly) different schemas.
+    ///
+    /// Because this feature tracks only the arity of the output, other layers
+    /// of MZ must ensure the correctness of this behavior, i.e. that the
+    /// schemas of the wider relation is a strict prefix of the narrower
+    /// versions.
+    pub arity: usize,
     /// The collection metadata needed to write the exported data
```

Comment:
this design is different than what was suggested in https://github.com/MaterializeInc/materialize/pull/25045#pullrequestreview-1877787332

can you explain the reasoning in choosing to use a single `arity` value rather than the proposed `projection: Vec<usize>` to select which columns in the upstream are required in this export?

Reply:
Sure, I can add more detail around that. Using `arity` rather than `Vec<usize>` is sufficient for the common use case of adding new columns (we require the old schema to be a strict prefix of the new schema). It's much, much simpler, more performance sensitive, and doesn't preclude doing something more exotic with the projections in the fullness of time.

Another way to think about this is that to support the common use case of adding columns from PG, the `projection` list would be `vec![0, 1, 2, 3]`, which reduces down to the concept of `arity` without the performance penalty of having to unpack the row.